File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
class Solution {
public:
    unordered_map<int,int> mapping;
    unordered_map<int,int> rank;
    void add(int x){
        if(!mapping.count(x)){
            mapping[x] = x;
            rank[x]=1;
        }
    }
    void doUnion(int x, int y){
        auto X = findParent(x);
        auto Y = findParent(y);
        if(X==Y) return;
        if(rank[X]<rank[Y]) {
            mapping[X] = Y;
            rank[Y] += rank[X];
        }
        else{
            mapping[Y] = X;
            rank[X] += rank[Y];
        }
    }
    int findParent(int x){
        if(x==mapping[x]){
            return x;
        }
        else{
            return mapping[x] = findParent(mapping[x]);
        }
    }
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        mapping = unordered_map<int,int>();
        rank = unordered_map<int,int>();
        for(auto edge:edges){
            add(edge[0]);
            add(edge[1]);
            doUnion(edge[0],edge[1]);
        }
        
        return findParent(source)==findParent(destination);
    }
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        adj = defaultdict(list)
        def dfs(node,seen=set()):
            if node==destination:
                return True
            for nex in adj[node]:
                if nex not in seen:
                    seen.add(nex)
                    if dfs(nex,seen):
                        return True
            return False
        
        for u,v in edges:
            adj[u].append(v)
            adj[v].append(u)
        
        return dfs(source)
        